Renaming and extension of APIs

dmgav opened the pull request:
- Renamed APIs in following the discussed naming scheme.

- Corrected multiple bugs and inconsistencies in data and error handling.

- Extended test coverage.

- Improved stability of 0MQ communication between Web Server and RE Manager.

- Update documentation in `README.rst`.

Note, that some functions may return data that is arranged differently. E.g. `queue_plan_add` (`/queue/plan/add`) returns the added plan as the value of the key `plan` of a dictionary which also contains the keys `success` and `msg` (error message if any).

cryos commented on bluesky_queueserver/server/server.py:
```diff
@@ -102,140 +121,284 @@ class REPauseOptions(str, Enum):
 
-@app.post("/pop_from_queue")
-async def pop_from_queue_handler():
+@app.get("/queue/plan/get")
```
It is odd for a GET to have a body/payload, quoting 

> Server semantics for GET, however, are restricted such that a body, if any, has no semantic meaning to the request.

POSTs without bodies feels odd, but GET with a body is even odder, I would expect this to be a POST, or to simply encode in the URL, the [HTTP 1.1 spec says](https://tools.ietf.org/html/rfc2616#section-9.3):

> The GET method means retrieve whatever information ([...]) is identified by the Request-URI.

cryos replied:
Can plans not just be retrieved using simple index or identifiers? So `/queue/plan/get/2` for example would get the second plan.

dmgav replied:
I can change it to POST (as remove). My idea was to keep the call to `/queue/plan/get` identical to the call for `/queue/plan/remove' method. Remove method calls 'queue_plan_get' underneath anyway, because items in the queue are not naturally accessed by index: you can pop a given plan, not plan at a given location. I was also thinking of extending the structure of parameters.

cryos replied:
I think making it POST is far less surprising than a GET that requires a body. I have used many APIs where POST was used on something that doesn't change data on the server but needs to support more complex query parameters than can reasonably be encoded in the URL.

cryos reviewed (commented):
I find a GET request with a body odd, and wonder if the stop should have a body for canceling, and I think cancel would be clearer than deactivate stop

dmgav commented on README.rst:
```diff
@@ -224,6 +224,16 @@ Start execution of the plan queue. The environment MUST be opened before queue c
+  qserver -c queue_stop_deactivate
+
+  http POST http://localhost:8080/queue/stop/activate
+  http POST http://localhost:8080/queue/stop/deactivate
```
At this point I keep command line in sync with HTTP request, but it could be polished later to be more user friendly when the set of APIs settles.

cryos replied:
Yes, `/queue/stop` and `/queue/stop/cancel` was my other thought too. I think it looks more natural.

dmgav commented:
So I changed GET to POST for `/queue/plan/get`. I also replaced `/queue/stop/activate` with `/queue/stop` and `/queue/stop/deactivate` with `/queue/stop/cancel`.

cryos reviewed (approved):
LGTM, I need to start integrating with more of the API, but it seems it would be better to do that after this is merged. Thanks for all the work on the RESTful side of the API!

mrakitin reviewed (commented):
That's a significant rework of the previous API to standardize it to what we discussed. I like the documentation listing both the CLI commands and the corresponding http requests. I have a few non-blocking suggestions, otherwise looks great!

dmgav commented on bluesky_queueserver/server/server.py:
```diff
@@ -102,140 +121,284 @@ class REPauseOptions(str, Enum):
+
+
+@app.post("/test/manager/kill")
+async def test_manager_kill_handler():
```
I think it has to be in a file with a name that starts with `test`.

mrakitin replied:
Yes, indeed.

dmgav commented on bluesky_queueserver/server/server.py:
```diff
@@ -102,140 +121,284 @@ class REPauseOptions(str, Enum):
+            extra_keys.add(key)
+        else:
+            if key in r_keys:
+                r_keys.remove(key)
```
Done

dmgav commented on bluesky_queueserver/server/server.py:
```diff
@@ -102,140 +121,284 @@ class REPauseOptions(str, Enum):
+async def queue_stop_cancel():
     """
-    msg = await re_server.send_command(command="list_allowed_plans_and_devices")
+    Deactivate the sequence of stopping the queue.
```
Done

dmgav commented:
I think I addressed all the issues, so I will merge the PR.
